### PR TITLE
Support drawing rounded rect backgrounds behind TopBar buttons

### PR DESCRIPTION
**Pull request by yogevbd**
* Add `button.iconBackground` option with:
  * `color`
  * `disabledColor`
  * `cornerRadius`
  * `width`
  * `height`
- [x] iOS
- [x] Android
- [x] Documentation

* Refactor buttons creation flow on iOS.

### CODE REVIEW

**Comment by mrousavy** [comment]
Does this use `RCTText` or are you manually applying styles?

**Comment by yogevbd** [comment]
@mrousavy it's not related to the title text, only to the icon background.

**Review by guyca — CHANGES_REQUESTED** [review]

File: lib/android/app/src/main/java/com/reactnativenavigation/views/stack/topbar/titlebar/IconBackgroundDrawable.kt [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    private val cornerRadius: Float =  UiUtils.dpToPx(context, iconBackgroundOptions.cornerRadius.get(0.0).toFloat())
+
+    override fun draw(canvas: Canvas) {
+        drawPath(canvas)
+        drawBackgroundColor(canvas)
+        drawBitmap(canvas)
+    }
+
+    override fun getIntrinsicWidth(): Int {
+        return UiUtils.dpToPx(context, getBackgroundWidth())
```
[the remark this answers is not in the document]
Comment (yogevbd) [reply lib/android/app/src/main/java/com/reactnativenavigation/views/stack/topbar/titlebar/IconBackgroundDrawable.kt]: Removed this function as it's no longer needed

File: lib/android/app/src/main/java/com/reactnativenavigation/views/stack/topbar/titlebar/IconBackgroundDrawable.kt [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    override fun getIntrinsicHeight(): Int {
```
[the remark this answers is not in the document]
Comment (yogevbd) [reply lib/android/app/src/main/java/com/reactnativenavigation/views/stack/topbar/titlebar/IconBackgroundDrawable.kt]: Removed this function as it's no longer needed

File: lib/ios/RNNButtonsPresenter.m [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    for (UIBarButtonItem *buttonItem in removedButtons) {
+        RNNReactView *reactView = buttonItem.customView;
+        if ([reactView isKindOfClass:[RNNReactView class]]) {
+            [reactView componentDidDisappear];
+            [_componentRegistry removeChildComponent:reactView.componentId];
+        }
+    }
+}
+
+- (RNNUIBarButtonItem *)buildButton:(RNNButtonOptions *)button {
```
[the remark this answers is not in the document]
Comment (yogevbd) [reply lib/ios/RNNButtonsPresenter.m]: done

File: lib/android/app/src/main/java/com/reactnativenavigation/options/IconBackgroundOptions.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    public IconBackgroundOptions copy() {
+        IconBackgroundOptions options = new IconBackgroundOptions();
+        options.mergeWith(this);
+        return options;
+    }
+
+    public boolean hasValue() {
+        return color.hasValue() ||
```
[the remark this answers is not in the document]
Comment (yogevbd) [reply lib/android/app/src/main/java/com/reactnativenavigation/options/IconBackgroundOptions.java]: Color should be mandatory, width and height can have the default icon size

File: lib/ios/RNNButtonBuilder.m [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        @throw [NSException
+            exceptionWithName:@"NSInvalidArgumentException"
+                       reason:[@"button id is not specified "
+                                  stringByAppendingString:[button.text getWithDefaultValue:@""]]
+                     userInfo:nil];
+    }
+}
+
+- (RNNUIBarButtonItem *)build:(RNNButtonOptions *)button onPress:(RNNButtonPressCallback)onPress {
+    [self assertButtonId:button];
```
Comment (guyca) [inline lib/ios/RNNButtonBuilder.m:32]: Typically, a caller function should be placed above the callee. High-level functions are placed above low-level functions.

File: lib/ios/RNNButtonBuilder.m [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+@implementation RNNButtonBuilder
+
+- (instancetype)initWithComponentRegistry:(id)componentRegistry {
+    self = [super init];
+    self.componentRegistry = componentRegistry;
+    return self;
+}
+
+- (void)bindViewController:(UIViewController<RNNLayoutProtocol> *)viewController {
```
Comment (guyca) [inline lib/ios/RNNButtonBuilder.m:17]: The builder pattern is used in direct substitution to a class' constructor, therefore all parameters the constructor requires should be available. Using a method prefixed with `bind` sounds foreign here.

Let's rename it to `setViewController` and have the method return the builder, this way we can chain its invocation with the `build()` method.

**Review by guyca — CHANGES_REQUESTED** [review]

File: lib/android/app/src/main/java/com/reactnativenavigation/options/IconBackgroundOptions.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    public IconBackgroundOptions copy() {
+        IconBackgroundOptions options = new IconBackgroundOptions();
+        options.mergeWith(this);
+        return options;
+    }
+
+    public boolean hasValue() {
+        return color.hasValue() &&
```
[the remark this answers is not in the document]
Comment (yogevbd) [reply lib/android/app/src/main/java/com/reactnativenavigation/options/IconBackgroundOptions.java:46]: No, it falls back to the original icon size

**Review by guyca — APPROVED** [review]

**Pull request merged**